### PR TITLE
issue #95 - add salt to md5 and email anonymizers

### PR DESCRIPTION
**Pull request by pounard**

### CODE REVIEW

File: docs/content/anonymization/core-anonymizers.md [hunk 1]
```diff
@@ -254,6 +262,17 @@ customer:
 ```
 :::
 
+:::tip
+Hashing a string is not anonymizing it because hash functions have a reproducible output,
+this in order to avoid decrypting data using rainbow tables, per default a salt will be
+added to string values prior to hasing.
```
Comment (Lonnytunes) [inline docs/content/anonymization/core-anonymizers.md:268]:
```suggestion
Hashing a string is not anonymizing it because hash functions have a reproducible output.
In order to avoid decrypting data using rainbow tables, a salt will be added by default to
string values prior to hashing.
```

File: docs/content/anonymization/core-anonymizers.md [hunk 2]
```diff
@@ -81,6 +81,14 @@ customer:
 ```
 :::
 
+:::tip
+Username value is salted prior to be hashed using md5 in order to prevent reverse hashing
+with rainbow tables, but salt is global across the same anonymizer run, which keeps a single
+value when the same email is anonymized more than once.
```
[the remark this answers is not in the document]
Comment (pounard) [reply docs/content/anonymization/core-anonymizers.md:87]: Actually I changed "username" to "email" (will push that soon) it's more correct.

File: docs/content/backup_restore.md [hunk 3]
```diff
@@ -32,10 +32,10 @@ console db-tools:backup --connection other_connection_name
 
 ### Excluded tables
 
-You may have configured [tables to be excluded in the bundle configuration](./configuration#excluded-tables).
+You may have configured [tables to be exclude in the bundle configuration](./configuration#excluded-tables).
```
Comment (Lonnytunes) [inline docs/content/backup_restore.md:35]:
```suggestion
You may have configured [tables to exclude in the bundle configuration](./configuration#excluded-tables).
```

Comment (pounard) [reply docs/content/backup_restore.md:35]: Not my PR !

Comment (pounard) [reply docs/content/backup_restore.md]: Sorry, I forgot to rebase actual main branch. Fixed.

File: src/Anonymization/Anonymizer/Core/EmailAnonymizer.php [hunk 4]
```diff
@@ -14,6 +14,7 @@
     pack: 'core',
     description: <<<TXT
     Anonymize email addresses. You can choose a domain and a tld with option 'domain'.
+    Values are salted to avoid reverse hashing with option 'use_salt' (default true).
```
Comment (Lonnytunes) [inline src/Anonymization/Anonymizer/Core/EmailAnonymizer.php:17]:
```suggestion
    Values are salted to avoid reverse hashing with option 'use_salt' (default: true).
```

File: src/Anonymization/Anonymizer/Core/Md5Anonymizer.php [hunk 5]
```diff
@@ -11,7 +11,11 @@
 #[AsAnonymizer(
     name: 'md5',
     pack: 'core',
-    description: 'Anonymize a column by hashing its value.'
+    description: <<<TXT
+    Anonymize a column by hashing its value.
+    Options are 'use_salt' (default true).
+    Using a salt prevents reverse hash of values using rainbow tables.
```
Comment (Lonnytunes) [inline src/Anonymization/Anonymizer/Core/Md5Anonymizer.php:17]: I think "... prevents reversing (the) hash of ..." or "... prevents reverse hashing of ..." would be a more correct wording.

File: src/Anonymization/Anonymizator.php [hunk 6]
```diff
@@ -52,12 +58,17 @@ protected function createAnonymizer(AnonymizerConfig $config): AbstractAnonymize
...
             $config->table,
             $config->targetName,
             $this->connection,
             $config->options,
         );
+        \assert($ret instanceof AbstractAnonymizer);
+
+        $ret->setSalt($this->getSalt());
+
+        return $ret;
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizator.php:71]: Every `Anonymizer` has to implement `AbstractAnonymizer`, so we could just do

```suggestion
        return new $className(
            $config->table,
            $config->targetName,
            $this->connection,
            $config->options,
            $this->getSalt(),
        );
        return $ret;
```

no ?

File: src/Anonymization/Anonymizer/AbstractAnonymizer.php [hunk 7]
```diff
@@ -27,8 +27,14 @@ final public function __construct(
         protected string $columnName,
         protected Connection $connection,
         protected Options $options,
+        private ?string $salt = null,
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizer/AbstractAnonymizer.php:30]: Why not 
```suggestion
        private string $salt,
```
?

And put the `::generateRandomSalt` method on the `Anonymizator` class ?

File: docs/content/anonymization/core-anonymizers.md [hunk 8]
```diff
@@ -254,6 +262,17 @@ customer:
 ```
 :::
 
+:::tip
```
[the remark this answers is not in the document]
Comment (pounard) [reply docs/content/anonymization/core-anonymizers.md:265]: As you wish, both are fine because it also document how to deactivate, but let's go for info.

File: src/Anonymization/Anonymizator.php (hunk 6 above)
Comment (pounard) [reply src/Anonymization/Anonymizator.php:71]: Yes, indeed, I'll change that.

File: src/Anonymization/Anonymizer/AbstractAnonymizer.php (hunk 7 above)
Comment (pounard) [reply src/Anonymization/Anonymizer/AbstractAnonymizer.php:30]: I'm not sure, anonymizers could be services one day, and I don't see why we should then force the salt to be in the constructor ?

Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/AbstractAnonymizer.php:30]: makes sens, then maybe we should push it throw options ?

Comment (pounard) [reply src/Anonymization/Anonymizer/AbstractAnonymizer.php:30]: If the correct word was "thought" then yes, it'd also make sense, as you prefer I have no strong opinion. Maybe we'd need some kind of common validation callback for options, that would validate all common options ?

**Comment by pounard** [comment]
Ah ! Interesting, since I add the string concat, some RDBMS seem to ignore that NULL values are NULL. I need to add a `$this->getSetIfNotNullExpression()` call.

File: src/Anonymization/Anonymizer/AbstractAnonymizer.php (hunk 7 above)
Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/AbstractAnonymizer.php:30]: One of the goal of this class was to not have to update the AbstractAnonymizer constructor signature in the future and as this property will only affect a few anonymizers, I prefer that it not pass the options.

Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/AbstractAnonymizer.php:30]: And yes, it could be a good idea to had a method `AbstractAnonymizer::validateOptions(Options $options)`  to let each anonymizer validate the options you give to it

File: src/Anonymization/Anonymizer/Core/Md5Anonymizer.php [hunk 9]
```diff
@@ -21,10 +25,21 @@ class Md5Anonymizer extends AbstractAnonymizer
...
+        if ($this->options->get('use_salt', true)) {
+            $columnExpr = $expr->concat($columnExpr, $expr->value($this->getSalt()));
+
+            $update->set(
+                $this->columnName,
+                // Work around some RDBMS not seeing the NULL value anymore
+                // once we added the string concat.
+                $this->getSetIfNotNullExpression(
+                    $expr->md5($columnExpr)
+                ),
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizer/Core/Md5Anonymizer.php:39]: why is this needed here and not on the EmailAnonymizer ? :thinking:

File: src/Anonymization/Anonymizer/AbstractAnonymizer.php (hunk 7 above)
Comment (SimonMellerin) [reply src/Anonymization/Anonymizer/AbstractAnonymizer.php:30]: yep

File: src/Anonymization/Anonymizer/Core/Md5Anonymizer.php (hunk 9 above)
Comment (pounard) [reply src/Anonymization/Anonymizer/Core/Md5Anonymizer.php:39]: It is already in email anonymizer if I remember correctly.

Comment (pounard) [reply src/Anonymization/Anonymizer/Core/Md5Anonymizer.php:39]: Yes, I can confirm that.

**Review by SimonMellerin — APPROVED** [review]
